python/ray/autoscaler/commands.py

```diff
@@ -344,15 +344,47 @@ def remaining_nodes():
 
             return head + workers
 
+        def try_docker_stop(node):
+            container_name = config.get("docker", {}).get("container_name")
+            if container_name is None:
+                return
+            try:
+                updater = NodeUpdaterThread(
+                    node_id=node,
+                    provider_config=config["provider"],
+                    provider=provider,
+                    auth_config=config["auth"],
+                    cluster_name=config["cluster_name"],
+                    file_mounts=config["file_mounts"],
+                    initialization_commands=[],
+                    setup_commands=[],
+                    ray_start_commands=[],
+                    runtime_hash="",
+                    file_mounts_contents_hash="",
+                    is_head_node=False,
+                    docker_config=config.get("docker"))
+                _exec(
+                    updater,
+                    f"docker stop {container_name}",
+                    False,
+                    False,
+                    run_env="host")
+            except Exception:
+                cli_logger.warning(f"Docker stop failed on {node}")
+                cli_logger.old_warning(f"Docker stopped on {node}")
+
         # Loop here to check that both the head and worker nodes are actually
         #   really gone
         A = remaining_nodes()
+
+        for node in A:
```

Comment:
Can we do this loop only in "docker mode"?

Reply:
Sure--I moved the check to outside the loop!